### PR TITLE
feat(datasource/pod): Add GitHub enterprise support for pods

### LINKED ISSUE #6568: CocoaPods GitHub enterprise hosted dependency support
**What Renovate type are you using?**
We are currently using renovate-on-prem for our web app

**Describe the bug**
We are looking to use the same in iOS, automate dependency updates for Cocoapods. We noticed an issue with GitHub enterprise hosted dependency (private Cocoapods repo) as they are not updated. Does renovate support CocoaPods GitHub enterprise hosted dependency? Any suggestion or fix for this issue?

**Did you see anything helpful in debug logs?**
`"msg":"Failed to look up dependency ___________"`

**To Reproduce**
1.  Create a private CocoaPods repo on github enterprise (github enterprise and requires auth for read access perhaps)
2. Create a sample iOS project with CocoaPods that has pod dependency that's hosted in private CocoaPods repo on github enterprise
3. Run renovate and notice the error to resolve dependency

**Additional context**

### PR DESCRIPTION
**Pull request by asoneji**
## Changes:

Add GitHub enterprise support for pods

## Context:

fixes #6568
massage URL moved to PR #12899

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/metadata.ts [hunk 1]
```diff
@@ -110,8 +110,14 @@ const manualSourceUrls = {
...
+  let massagedUrl = url;
+
+  if (url.startsWith('git@')) {
+    massagedUrl = url.replace(':', '/').replace('git@', 'https://');
+  }
+
+  return massagedUrl
     .replace('http:', 'https:')
     .replace('http+git:', 'https:')
     .replace('https+git:', 'https:')
```
Comment (asoneji) [inline lib/datasource/metadata.ts:123]: Thanks, @ranolf for the suggestion:

```suggestion
    .replace('https+git:', 'https:')
    .replace('ssh://git@','https://')
```

File: lib/datasource/pod/index.ts [hunk 2]
```diff
@@ -32,13 +33,15 @@ function shardParts(lookupName: string): string[] {
 
 function releasesGithubUrl(
   lookupName: string,
-  opts: { account: string; repo: string; useShard: boolean }
+  opts: { hostURL: string; account: string; repo: string; useShard: boolean }
 ): string {
-  const { useShard, account, repo } = opts;
-  const prefix = 'https://api.github.com/repos';
+  const { hostURL, account, repo, useShard } = opts;
+  const prefix = hostURL
```
Comment (asoneji) [inline lib/datasource/pod/index.ts:39]: Thanks, @ranolf for the suggestion:

```suggestion
  const prefix = hostURL && !hostURL.includes('api.github.com')
```

File: lib/datasource/pod/index.ts [hunk 3]
```diff
@@ -98,14 +101,17 @@ const githubRegex = regEx(
   /^https:\/\/github\.com\/(?<account>[^/]+)\/(?<repo>[^/]+?)(\.git|\/.*)?$/
 );
 
+const githubEnterpriseRegex = regEx(
+  /(?<hostURL>(^https:\/\/github\..*\.com))\/(?<account>[^/]+)\/(?<repo>[^/]+?)(\.git|\/.*)?$/
```
Comment (asoneji) [inline lib/datasource/pod/index.ts:105]: Thanks, @ranolf for the suggestion:

```suggestion
  /(?<hostURL>(^https:\/\/github\..+\.[a-zA-z0-9-]+))\/(?<account>[^/]+)\/(?<repo>[^/]+?)(\.git|\/.*)?$/
```

File: lib/datasource/pod/index.ts [hunk 4]
```diff
@@ -98,14 +101,17 @@ const githubRegex = regEx(
...
+);
+
 async function getReleasesFromGithub(
   lookupName: string,
-  registryUrl: string,
+  match: RegExpExecArray,
```
Comment (asoneji) [inline lib/datasource/pod/index.ts:110]: Thanks, @ranolf for the suggestion:

```suggestion
  opts: { hostURL: string; account: string; repo: string; useShard: boolean }
```

File: lib/datasource/metadata.ts [hunk 5]
```diff
@@ -110,11 +110,18 @@ const manualSourceUrls = {
...
 const gitPrefix = regEx('^git:/?/?');
 
-function massageGithubUrl(url: string): string {
-  return url
+export function massageGithubUrl(url: string): string {
+  let massagedUrl = url;
+
+  if (url.startsWith('git@')) {
+    massagedUrl = url.replace(':', '/').replace('git@', 'https://');
+  }
```
Comment (viceice) [inline lib/datasource/metadata.ts:118]: Please ad a test for this

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/metadata.ts (hunk 5 above)
Comment (asoneji) [reply lib/datasource/metadata.ts:118]: Will do.

File: lib/datasource/pod/index.ts [hunk 6]
```diff
@@ -32,13 +33,15 @@ function shardParts(lookupName: string): string[] {
...
-  const { useShard, account, repo } = opts;
-  const prefix = 'https://api.github.com/repos';
+  const { hostURL, account, repo, useShard } = opts;
+  const prefix = hostURL
+    ? `${hostURL}/api/v3/repos`
+    : 'https://api.github.com/repos';
   const shard = shardParts(lookupName).join('/');
-  const suffix = useShard ? `${shard}/${lookupName}` : lookupName;
-  return `${prefix}/${account}/${repo}/contents/Specs/${suffix}`;
+  const suffix = useShard ? `${repo}/${shard}/${lookupName}` : lookupName;
```
Comment (asoneji) [inline lib/datasource/pod/index.ts:43]: Repo name (`Specs`) should only be appended again for shard URL only. Update the unit test and snapshot to reflect the same.

File: lib/datasource/metadata.ts [hunk 7]
```diff
@@ -110,11 +110,18 @@ const manualSourceUrls = {
 const githubPages = regEx('^https://([^.]+).github.com/([^/]+)$');
 const gitPrefix = regEx('^git:/?/?');
 
-function massageGithubUrl(url: string): string {
-  return url
+export function massageGithubUrl(url: string): string {
```
Comment (viceice) [inline lib/datasource/metadata.ts:113]: Can you extract this change to a separate pr please.

File: lib/datasource/pod/index.ts (hunk 6 above)
Comment (viceice) [reply lib/datasource/pod/index.ts:43]: Please provide a sample test repo, which shows this required changes.

**Review by viceice — COMMENTED** [review]
Looks good to me, but like to see some real tests, so we don't break something

File: lib/datasource/pod/index.ts (hunk 6 above)
Comment (asoneji) [reply lib/datasource/pod/index.ts:43]: * [Steps](https://guides.cocoapods.org/making/private-cocoapods.html) for creating repo.
* Creating a New PodSpec Repository [article1](https://medium.com/@shahabejaz/create-and-distribute-private-libraries-with-cocoapods-5b6507b57a03) and [article2](https://medium.com/keove/creating-private-public-pods-for-cocoapods-2fc69b30270b)
* Sample of repos
   * https://github.com/keove/PodSpecs
   * https://github.com/shahabejaz/PrivatePods
* Sample of repo with shared
   * https://github.com/CocoaPods/Specs/tree/master/Specs/a/7/6/Analytics

File: lib/datasource/metadata.ts (hunk 7 above)
Comment (asoneji) [reply lib/datasource/metadata.ts:113]: Sure created https://github.com/renovatebot/renovate/pull/12899

**Comment by asoneji** [comment]
> Looks good to me, but like to see some real tests, so we don't break something

I have tested this locally for repo with shard and without and added the same in the unit test.

**Review by viceice — DISMISSED** [review]
I don't use pods, so 🤞

**Comment by asoneji** [comment]
After chatting with @ranolf we updated the PR to support current and legacy pod repo URL format.

**Comment by asoneji** [comment]
@viceice @rarkins @JamieMagee  Can I please get help reviewing this PR again and get it merged. Thanks.

File: lib/datasource/pod/index.spec.ts [hunk 8]
```diff
@@ -56,6 +61,27 @@ describe('datasource/pod/index', () => {
...
+        .reply(404);
+      const res = await getPkgReleases({
+        ...config,
+        registryUrls: [
+          ...config.registryUrls,
+          'https://github.foo.com/foo/bar',
+        ],
+      });
+      expect(res).toBeNull();
+      expect(httpMock.getTrace()).toMatchSnapshot();
```
[the remark this answers is not in the document]
Comment (asoneji) [reply lib/datasource/pod/index.spec.ts:83]: The expect for snapshot matching was added as the existing test had them. Should this be removed for all the test in `lib/datasource/pod/index.spec.ts` or just the new tests added?

File: lib/datasource/pod/index.ts [hunk 9]
```diff
@@ -21,6 +22,13 @@ const cacheMinutes = 30;
 const githubHttp = new GithubHttp(id);
 const http = new Http(id);
 
+const enum URLFormatOptions {
+  WithSharedWithSpec = 'WITH_SHARD_WITH_SPEC',
+  WithSharedWithoutSpec = 'WITH_SHARD_WITHOUT_SPEC',
+  WithSpecsWithoutShard = 'WITH_SPECS_WITHOUT_SHARD',
+  WithoutSpecsWithoutShard = 'WITHOUT_SPECS_WITHOUT_SHARD',
```
[the remark this answers is not in the document]
Comment (asoneji) [reply lib/datasource/pod/index.ts]: updated 0e16403bf

File: lib/datasource/pod/index.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (viceice) [reply lib/datasource/pod/index.spec.ts:83]: remove, they are no longer nesessary.

File: lib/datasource/pod/index.ts [hunk 10]
```diff
@@ -21,6 +22,13 @@ const cacheMinutes = 30;
 const githubHttp = new GithubHttp(id);
 const http = new Http(id);
 
+const enum URLFormatOptions {
+  WithSharedWithSpec = 0,
+  WithSharedWithoutSpec = 1,
+  WithSpecsWithoutShard = 2,
+  WithoutSpecsWithoutShard = 3,
```
Comment (viceice) [inline lib/datasource/pod/index.ts:29]:
```suggestion
  WithSharedWithSpec,
  WithSharedWithoutSpec,
  WithSpecsWithoutShard,
  WithoutSpecsWithoutShard,
```

I think this will work too and use automatical numbering. 😉

Comment (asoneji) [reply lib/datasource/pod/index.ts]: Ahh thanks for the suggestion.

File: lib/datasource/pod/index.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (asoneji) [reply lib/datasource/pod/index.spec.ts]: Removed unnecessary snapshot matching 772cb24c0.

File: lib/datasource/pod/index.ts [hunk 11]
```diff
@@ -98,26 +119,54 @@ const githubRegex = regEx(
...
+);
+
 async function getReleasesFromGithub(
   lookupName: string,
-  registryUrl: string,
-  useShard = false
+  opts: { hostURL: string; account: string; repo: string },
+  useShard = true,
+  useSpecs = true,
+  urtFormatOptions = URLFormatOptions.WithSharedWithSpec
```
Comment (viceice) [inline lib/datasource/pod/index.ts:131]: is this somehow configurable? Or are you only using the default value? Are you planning more changes?

Comment (asoneji) [reply lib/datasource/pod/index.ts:131]: the `getReleasesFromGithub` method tries all the different url format using this configuration and I am using this as the default value. I am not planning to add more changes but in future there could be another format someone else may require.

**Review by viceice — COMMENTED** [review]
As we now do more requests, there should be some caching, so we can remember the final url.

How does the normal pod resolve work?

Like to have comments of other maintainers too, as i don't know / use pods

**Comment by asoneji** [comment]
> As we now do more requests, there should be some caching, so we can remember the final URL.

Do we have examples of other datasource that I should refer to for caching? Could this be a follow-up PR?

> How does the normal pod resolve work?

We try 4 different formats of URL but the first one we try which is URL with `Specs` in path and the shard URL format should resolve for the normal pod.

> Like to have comments of other maintainers too, as i don't know / use pods

@rarkins @JamieMagee could you please provide your review. thanks.

File: lib/datasource/pod/index.ts [hunk 12]
```diff
@@ -32,13 +40,26 @@ function shardParts(lookupName: string): string[] {
 
 function releasesGithubUrl(
   lookupName: string,
-  opts: { account: string; repo: string; useShard: boolean }
+  opts: {
+    hostURL: string;
```
Comment (JamieMagee) [inline lib/datasource/pod/index.ts:44]:
```suggestion
    hostURL?: string;
```

From the codepath below, it looks like it's optional?

File: lib/datasource/pod/index.ts [hunk 13]
```diff
@@ -32,13 +40,26 @@ function shardParts(lookupName: string): string[] {
...
+    account: string;
+    repo: string;
+    useShard: boolean;
+    useSpecs: boolean;
+  }
```
Comment (JamieMagee) [inline lib/datasource/pod/index.ts:49]: This is a bit of a nitpick, but what do you think about extracting this complex type?

**Comment by JamieMagee** [comment]
> Do we have examples of other datasource that I should refer to for caching? Could this be a follow-up PR?

I can take care of caching in a follow-up PR. I've been making my way through all the other datasources.

File: lib/datasource/pod/index.ts [hunk 14]
```diff
@@ -98,26 +119,54 @@ const githubRegex = regEx(
...
+);
+
 async function getReleasesFromGithub(
   lookupName: string,
-  registryUrl: string,
-  useShard = false
+  opts: { hostURL: string; account: string; repo: string },
+  useShard = true,
+  useSpecs = true,
+  urtFormatOptions = URLFormatOptions.WithShardWithSpec
```
Comment (JamieMagee) [inline lib/datasource/pod/index.ts:131]:
```suggestion
  urlFormatOptions = URLFormatOptions.WithShardWithSpec
```

File: lib/datasource/pod/index.ts (hunk 13 above)
Comment (asoneji) [reply lib/datasource/pod/index.ts]: I am ok extracting the options from the complex type and passing them as parameters. I was just following the current method which already had `opts: { account: string; repo: string; useShard: boolean }`. Maybe be what I should do is pull out hostURL and add useSpecs to the options. so it will look like this:

```TypeScript
function releasesGithubUrl(
  lookupName: string,
  opts: {
    account: string;
    repo: string;
    useShard: boolean;
    useSpecs: boolean;
  },
  hostURL?: string,
):
```

File: lib/datasource/pod/index.ts [hunk 15]
```diff
@@ -98,26 +119,54 @@ const githubRegex = regEx(
   /^https:\/\/github\.com\/(?<account>[^/]+)\/(?<repo>[^/]+?)(\.git|\/.*)?$/
 );
 
+const githubEnterpriseRegex = regEx(
+  /(?<hostURL>(^https:\/\/github\..+\.[a-zA-z0-9-]+))\/(?<account>[^/]+)\/(?<repo>[^/]+?)(\.git|\/.*)?$/
```
Comment (JamieMagee) [inline lib/datasource/pod/index.ts:123]: This regex requires GitHub Enterprise to be under `github.foo.bar`, but I've seen `gh.foo.bar`, `ghe.foo.bar`, etc. Really all you need is something that is an absolute URL and ends in `/account/repo`.

File: lib/datasource/pod/index.ts (hunk 14 above)
Comment (asoneji) [reply lib/datasource/pod/index.ts]: thanks for finding the typo. Updated in 3594dd917.

File: lib/datasource/pod/index.ts (hunk 13 above)
Comment (JamieMagee) [reply lib/datasource/pod/index.ts]: Okay, let's leave it for now. It was a nitpick anyway 😄

File: lib/datasource/pod/index.ts (hunk 15 above)
Comment (asoneji) [reply lib/datasource/pod/index.ts]: Updated 178feba6d

Comment (asoneji) [reply lib/datasource/pod/index.ts]: Regex tester: https://regex101.com/r/3fo4dT/1

**Comment by asoneji** [comment]
@viceice @rarkins @JamieMagee Can I please get help reviewing this PR again and get it merged. Thanks.

File: lib/datasource/pod/index.ts [hunk 16]
```diff
@@ -95,29 +117,53 @@ async function requestGithub<T = unknown>(
 }
 
 const githubRegex = regEx(
-  /^https:\/\/github\.com\/(?<account>[^/]+)\/(?<repo>[^/]+?)(\.git|\/.*)?$/
+  /(?<hostURL>(^https:\/\/[a-zA-z0-9-.]+))\/(?<account>[^/]+)\/(?<repo>[^/]+?)(\.git|\/.*)?$/
```
Comment (viceice) [inline lib/datasource/pod/index.ts:120]: this would now also match `https://gitea.com/org/repo.git`, so it can probably fail for existing users now?

**Review by viceice — COMMENTED** [review]
Code is ok from my side, not sure about the pods handling. 🤷‍♂️

File: lib/datasource/pod/index.ts (hunk 16 above)
Comment (asoneji) [reply lib/datasource/pod/index.ts:120]: To confirm this locally I added the below tests:

```
    it('returns null for 404 Github 2', async () => {
      httpMock
        .scope('https://gitea.com')
        .get('/api/v3/repos/org/repo/contents/Specs/a/c/b/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/a/c/b/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/Specs/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/foo')
        .reply(404);
      const res = await getPkgReleases({
        ...config,
        registryUrls: [
          ...config.registryUrls,
          'https://gitea.com/org/repo',
        ],
      });
      expect(res).toBeNull();
    });



    it('processes real data from Github2 with shard with specs', async () => {
      httpMock
        .scope('https://gitea.com')
        .get('/api/v3/repos/org/repo/contents/Specs/a/c/b/foo')
        .reply(200, [{ name: '1.2.3' }]);
      const res = await getPkgReleases({
        ...config,
        registryUrls: ['https://gitea.com/org/repo'],
      });
      expect(res).toEqual({
        registryUrl: 'https://gitea.com/org/repo',
        releases: [
          {
            version: '1.2.3',
          },
        ],
      });
    });
    it('processes real data from Github2 with shard without specs', async () => {
      httpMock
        .scope('https://gitea.com')
        .get('/api/v3/repos/org/repo/contents/Specs/a/c/b/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/a/c/b/foo')
        .reply(200, [{ name: '1.2.3' }]);
      const res = await getPkgReleases({
        ...config,
        registryUrls: ['https://gitea.com/org/repo'],
      });
      expect(res).toEqual({
        registryUrl: 'https://gitea.com/org/repo',
        releases: [
          {
            version: '1.2.3',
          },
        ],
      });
    });
    it('processes real data from Github2 with specs without shard', async () => {
      httpMock
        .scope('https://gitea.com')
        .get('/api/v3/repos/org/repo/contents/Specs/a/c/b/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/a/c/b/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/Specs/foo')
        .reply(200, [{ name: '1.2.3' }]);
      const res = await getPkgReleases({
        ...config,
        registryUrls: ['https://gitea.com/org/repo'],
      });
      expect(res).toEqual({
        registryUrl: 'https://gitea.com/org/repo',
        releases: [
          {
            version: '1.2.3',
          },
        ],
      });
    });
    it('processes real data from Github2 without specs without shard', async () => {
      httpMock
        .scope('https://gitea.com')
        .get('/api/v3/repos/org/repo/contents/Specs/a/c/b/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/a/c/b/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/Specs/foo')
        .reply(404)
        .get('/api/v3/repos/org/repo/contents/foo')
        .reply(200, [{ name: '1.2.3' }]);
      const res = await getPkgReleases({
        ...config,
        registryUrls: ['https://gitea.com/org/repo'],
      });
      expect(res).toEqual({
        registryUrl: 'https://gitea.com/org/repo',
        releases: [
          {
            version: '1.2.3',
          },
        ],
      });
    });
  });

```

They all passed:
```
 PASS  lib/datasource/pod/index.spec.ts
  datasource/pod/index
    ✓ processes real data from Github Enterprise with shard with specs (4 ms)
    ✓ processes real data from Github Enterprise with shard without specs (9 ms)
    ✓ processes real data from Github Enterprise with specs without shard (10 ms)
    ✓ processes real data from Github Enterprise without specs without shard (15 ms)
    getReleases
      ✓ returns null for invalid inputs (24 ms)
      ✓ returns null for empty result (5 ms)
      ✓ returns null for 404 (18 ms)
      ✓ returns null for 404 Github enterprise (19 ms)
      ✓ returns null for 404 Github 2 (17 ms)
      ✓ returns null for 404 Github enterprise with different url style (15 ms)
      ✓ returns null for 401 (5 ms)
      ✓ throws for 429 (32 ms)
      ✓ returns null for unknown error (5 ms)
      ✓ processes real data from CDN (6 ms)
      ✓ processes real data from Github with shard with specs (5 ms)
      ✓ processes real data from Github with shard without specs (9 ms)
      ✓ processes real data from Github with specs without shard (12 ms)
      ✓ processes real data from Github without specs without shard (17 ms)
      ✓ processes real data from Github2 with shard with specs (5 ms)
      ✓ processes real data from Github2 with shard without specs (10 ms)
      ✓ processes real data from Github2 with specs without shard (13 ms)
      ✓ processes real data from Github2 without specs without shard (17 ms)

Test Suites: 1 passed, 1 total
Tests:       22 passed, 22 total
Snapshots:   0 total
Time:        3.498 s
Ran all test suites within paths "lib/datasource/pod/index.spec.ts".
```

Don't think we need to actually add these tests as we already have something similar.

**Comment by asoneji** [comment]
> Code is ok from my side, not sure about the pods handling. 🤷‍♂️

Pod handling I have tested locally and is working for both github and github enterprise after this change.

File: lib/datasource/pod/index.ts (hunk 16 above)
Comment (viceice) [reply lib/datasource/pod/index.ts:120]: yes, but gitea isn't github, so it will fail on runtime.

I think we should do an `HEAD` reuest to the api root and check for github specific headers. So we only try further requests on gihub compatible api.

Comment (asoneji) [reply lib/datasource/pod/index.ts:120]: If the host and path exist the `getReleasesFromGithub` method will be able to call the URL.

if the request passes it will return `releases`
else if the request fails it will return `null`

Isn't this sufficient?

Comment (asoneji) [reply lib/datasource/pod/index.ts:120]: We tried parsing the header but unfortunately in the error case it does not return the response headers and I would think it would be out of the scope of this PR to change the request and error handling approach. The impact of this is that we make extra few requests.

Comment (ranolf) [reply lib/datasource/pod/index.ts:120]: Fundamentally, the change here is isn't doing anything new: previously,  if the githubURL somehow matched it would make two requests, and the response and additional load would be 2 404'd requests.  With the new approach, if the githubURL that the user configures somehow matches, we'll now have 4 404'd requests.   We'd need to make 1 HTTP request  anyway just to check the headers.  

If a renovate user has configured the URL against gitea, it will still fail because it's not a github or github Enterprise server -- just as it does today.

Comment (asoneji) [reply lib/datasource/pod/index.ts:120]: @viceice I am resolving this comment...please unresolve it if you don't agree.

**Review by rarkins — APPROVED** [review]
